packages/taiko-client/driver/preconf_blocks/api.go

```diff
@@ -357,6 +378,127 @@ func (s *PreconfBlockAPIServer) BuildPreconfBlock(c echo.Context) error {
+			return s.returnError(c, http.StatusInternalServerError, err)
+		}
+		if preconfRouter == rpc.ZeroAddress {
+			log.Warn("Preconfirmation is disabled via taikoWrapper", "preconfRouter", preconfRouter.Hex())
+			return s.returnError(
+				c,
+				http.StatusInternalServerError,
+				errors.New("preconfirmation is disabled via taikoWrapper"),
+			)
+		}
+	}
+
+	// Check if the L2 execution engine is syncing from L1.
+	progress, err := s.rpc.L2ExecutionEngineSyncProgress(ctx, s.shastaIndexer.GetLastCoreState())
+	if err != nil {
+		return s.returnError(c, http.StatusBadRequest, err)
+	}
+	if progress.IsSyncing() {
+		return s.returnError(c, http.StatusBadRequest, errors.New("l2 execution engine is syncing"))
+	}
+
+	// Parse request body
+	reqBody := new(BuildPreconfBlockV2RequestBody)
+	if err := c.Bind(reqBody); err != nil {
+		return s.returnError(c, http.StatusUnprocessableEntity, err)
+	}
+	if reqBody.Preconfirmation == nil {
+		return s.returnError(c, http.StatusBadRequest, errors.New("preconfirmation is required"))
+	}
+	if len(reqBody.Transactions) == 0 {
+		return s.returnError(c, http.StatusBadRequest, errors.New("transactions (compressed txlist) is required"))
+	}
+
+	pc := reqBody.Preconfirmation.Commitment.Preconf
+	log.Info(
+		"🏗️ New preconfirmation commitment (V2)",
+		"blockNumber", pc.BlockNumber,
+		"anchorBlockNumber", pc.AnchorBlockNumber,
+		"rawTxListHash", pc.RawTxListHash,
+		"parentRawTxListHash", pc.ParentRawTxListHash,
+		"eop", pc.EOP,
+	)
+
+	// Store txlist preimage and publish to P2P if configured
+	if s.p2pNode != nil {
+		if err := s.publishPreconfirmationFromSidecar(ctx, reqBody.Preconfirmation, &ExecutableData{Transactions: reqBody.Transactions}, nil); err != nil {
+			log.Warn("Failed to publish preconfirmation commitment", "err", err)
+			// Fall through to local processing after storing preimage
+			s.txlistMu.Lock()
+			s.txlist[pc.RawTxListHash] = reqBody.Transactions
+			s.txlistMu.Unlock()
+		}
+	} else {
+		// No P2P configured; still store preimage for reconstruction
+		s.txlistMu.Lock()
+		s.txlist[pc.RawTxListHash] = reqBody.Transactions
+		s.txlistMu.Unlock()
+	}
+
+	// Process the commitment locally to reconstruct and insert the block.
+	// Note: OnUnsafePreconfirmationCommitment handles concurrency (locks internally).
+	var zeroPeer peer.ID
+	if err := s.OnUnsafePreconfirmationCommitment(ctx, zeroPeer, reqBody.Preconfirmation); err != nil {
+		return s.returnError(c, http.StatusInternalServerError, fmt.Errorf("failed to process preconfirmation commitment: %w", err))
+	}
+
+	// Fetch inserted header by number
+	header, err := s.rpc.L2.HeaderByNumber(ctx, pc.BlockNumber)
+	if err != nil || header == nil {
+		return s.returnError(c, http.StatusInternalServerError, fmt.Errorf("failed to fetch inserted header: %w", err))
+	}
+
+	// Update unsafe payload tracking
+	s.updateHighestUnsafeL2Payload(header.Number.Uint64())
+
+	// If EOP marker, update cache
+	if pc.EOP && s.rpc.L1Beacon != nil {
```

Comment:
i believe if we receive an EOP marker, we need to "advance the lookahead" here as well (ongoing discussions in partner channel), rather than wait for the pre-determined lookahead advancement.

Reply:
please check now